src/uct/ib/efa/ib_efa_md.c

```diff
@@ -0,0 +1,147 @@
+/**
+ * Copyright (c) 2021, NVIDIA CORPORATION. All rights reserved.
+ *
+ * See file LICENSE for terms.
+ */
+
+#ifdef HAVE_CONFIG_H
+#  include "config.h"
+#endif
+
+#include <uct/ib/efa/ib_efa.h>
+#include <uct/ib/base/ib_md.h>
+#include <ucs/type/status.h>
+#include <ucs/debug/log.h>
+
+
+typedef struct uct_ib_efa_mem {
+    uct_ib_mem_t        super;
+    uct_ib_mr_t         mrs[];
+} uct_ib_efa_mem_t;
+
+static uint64_t
+uct_ib_efadv_access_flags(const uct_ib_efadv_t *efadv)
+{
+    uint64_t access_flags = IBV_ACCESS_LOCAL_WRITE;
+#ifdef HAVE_DECL_EFA_DV_RDMA_READ
+    if (efadv->efadv_attr.device_caps & EFADV_DEVICE_ATTR_CAPS_RDMA_READ) {
+        access_flags |= IBV_ACCESS_REMOTE_READ;
+    }
+#endif
+    return access_flags;
+}
+
+static ucs_status_t uct_ib_efa_reg_key(uct_ib_md_t *md, void *address,
+                                       size_t length, uint64_t access_flags,
+                                       uct_ib_mem_t *ib_memh,
+                                       uct_ib_mr_type_t mr_type, int silent)
+{
+    uct_ib_efa_mem_t *memh = ucs_derived_of(ib_memh, uct_ib_efa_mem_t);
+
+    return uct_ib_reg_key_impl(md, address, length, access_flags,
+                               ib_memh, &memh->mrs[mr_type], mr_type, silent);
+}
+
+static ucs_status_t uct_ib_efa_dereg_key(uct_ib_md_t *md,
+                                         uct_ib_mem_t *ib_memh,
+                                         uct_ib_mr_type_t mr_type)
+{
+    uct_ib_efa_mem_t *memh = ucs_derived_of(ib_memh, uct_ib_efa_mem_t);
+
+    return uct_ib_dereg_mr(memh->mrs[mr_type].ib);
+}
+
+static uct_ib_md_ops_t uct_ib_efa_md_ops;
+
+static ucs_status_t uct_ib_efa_md_open(struct ibv_device *ibv_device,
+                                       const uct_ib_md_config_t *md_config,
+                                       uct_ib_md_t **p_md)
+{
+    ucs_status_t status;
+    uct_ib_device_t *dev;
+    uct_ib_efadv_md_t *md;
+    int num_mrs;
+
+    md = ucs_calloc(1, sizeof(*md), "ib_efa_md");
+    if (md == NULL) {
+        return UCS_ERR_NO_MEMORY;
+    }
+
+    dev = &md->super.dev;
+    dev->ibv_context = ibv_open_device(ibv_device);
+    if (dev->ibv_context == NULL) {
+        ucs_warn("ibv_open_device(%s) failed: %m", ibv_get_device_name(ibv_device));
+        status = UCS_ERR_IO_ERROR;
+        goto err;
+    }
+
+    md->super.config = md_config->ext;
+
+    status = uct_ib_device_query(dev, ibv_device);
+    if (status != UCS_OK) {
+        goto err_free_context;
+    }
```

Comment:
can extract common code with verbs?

Reply:
We certainly can (and should). Most of verbs md open code is also used for efa md open. Maybe I can extract the common code into something like `uct_ib_verbs_md_open_common()` that takes a `uct_ib_md_t *md` instead of allocating one.